### ISSUE TITLE
Support implicit named arguments in formatting macros

### ISSUE THREAD

**Issue by Undin** [intellij-rust/intellij-rust#7338]
RFC: https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html

Small example:
```rust
let (person, species, name) = ("Charlie Brown", "dog", "Snoopy");

// implicit named argument `person`
print!("Hello {person}");

// implicit named arguments `species` and `name`
format!("The {species}'s name is {name}.");
```

Note, `format_args_capture` is current feature name in rustc 1.54.0-nightly

* [ ] don't produce false-positives for name arguments #7332
* [ ] name resolution
* [ ] completion
* [ ] take into account in analysis (unused variables, for example)

**Comment by Kobzol** [comment]
I started working on name resolution, I'll send a PR soon.

**Comment by Kobzol** [comment]
@Undin I wanted to provide a reference provider that would resolve named arguments:
```rust
println!("{foo}", foo = 1); // clicking on {foo} would resolve to `foo =`

let bar = 1;
println!("{bar}"); // clicking on {bar} would resolve to `let bar`.
```
I started by refactoring the code from the format macro annotator, which basically parses the formatting string with a regex and then applies highlighting to it. But now that I see all of the things that could be supported (name resolution, completion, dataflow analysis, unresolved reference etc.), I wonder if it would be better to actually parse the formatting string into proper PSI elements, so that we can better provide these features?

So that e.g. in `println!("{foo}", foo = 1);` the argument `foo = 1` would be a named PSI element and `{foo}` would be another PSI element that would resolve to it. Maybe resolving to named arguments is overkill, but even for resolving to local bindings for implicit named arguments it could be easier with proper PSI parsing.

Is that the proper solution? Or should I stick to regex-based handling of the format string?

**Comment by baadc0de** [comment]
if `#![feature(format_args_capture)]` the 1.58.0 compiler now emits a warning:

> warning: the feature `format_args_capture` has been stable since 1.58.0 and no longer requires an attribute to enable

intellij-rust should detect that `#![feature(format_args_capture)]` is not needed anymore.

**Comment by kolloch** [comment]
Is there a workaround for rust `stable` on `1.58.x` to make intellij aware that format_args_capture is allowed? Thank you.

**Comment by Undin** [comment]
Actually, PSI implementation looks more promising because it allows:
- not use custom approaches during any analysis (from name resolution and completion and to `RsLivenessInspection` inspection) and just use psi for it
- using psi viewers tools (like `PsiViewer` plugin) to check psi structure without debugging
- parse complex expressions inside format string like `println!("{x.y + z.foo()}")` that's impossible with regex. Of course, such syntax is not allowed right now but I won't be surprised if it will be implemented eventually

But it **may** require a lot of changes in different parts of the plugin

**Comment by decatur** [comment]
Note that according RFC: https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html:
Downstream macros based on format_args! include but are not limited to format!, print!, write!, panic!, and macros in the log crate.
Currently, for example `debug!("{foo}")` with debug from the tracing crate will produce an error in intellij-rust.

**Comment by Kobzol** [comment]
For me, this does not produce an error/warning in the plugin:
```rust
fn main() {
    let x = 5;
    tracing::debug!("{x}");
}
```

**Comment by rsalmei** [comment]
It was fixed already, two releases ago I think.
It's just the highlighting missing now (and refactorings of course).

**Comment by mati865** [comment]
It also doesn't work in assertions like mentioned it https://github.com/intellij-rust/intellij-rust/issues/7087#issuecomment-1311536882

**Comment by flymonkey00** [comment]
Since rust clippy 1.66
https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args
inline format args becomes a recommended way for the format strings, but use this style will break current intellij rust plugin analysing.

**Comment by ozkriff** [comment]
The clippy's warning got temporarily downgraded https://github.com/rust-lang/rust-clippy/issues/10087 but it'll be enabled again when RA implements this feature.
